### PR TITLE
Extend and export printpoly

### LINKED ISSUE #77: Add support for different `show`s
### Description

Looking at the current `show`, I think it does an amazing job.

Do you think we could extend `show` to have

- support for printing the polynomial in the reverse order, and/or,
- printing the polynomial in its factored form?

### Motivation

We are working on a control systems toolbox, and we are requiring `Polynomials` in that toolbox. Generally, people in the control community prefer seeing the polynomials

- in descending exponents, if the system model is represented by transfer functions, or,
- in factored form such as `(s+r1)(s+r2)` for real roots (with corresponding multiplicities), or, `(s^2+as+b)` for complex conjugate roots, if the system is represented by zero-pole-gain models.

As a side note, in transfer function models, we are using a subset of the polynomials: polynomials with real coefficients. As a result, we can also have an efficient search strategy for `Poly{T<:Real}`s  when finding the complex conjugate pairs.

### Some Suggestions

We can have this behaviour (if that would be implemented in `Polynomials`), by

- either proper method overloads, or,
- maybe through a global setting in the package (maybe a `const Dict{Symbol,Any}` with a setting symbol pointing to a value type such as `Show{:Factored}`, `Show{:Descending}` and/or `Show{:Ascending}` to dispatch the corresponding `print` method for `show`, for example.

### PR DESCRIPTION
**Pull request by martinholters**
Make `printpoly` part of the public API and add options to reverse the term order and to combine the powers with a power already present in the variable (if any).

Closes #77.

### CODE REVIEW

File: REQUIRE [hunk 1]
```diff
@@ -1,2 +1,2 @@
 julia 0.4
-Compat 0.9.4
+Compat 0.9.5
```
Comment (martinholters) [inline REQUIRE]: This is only needed for the test (for `String(take!(buf))`), but introducing a separate `test/REQUIRE` file just for this seemed a bit silly.

**Comment by jverzani** [comment]
Looks great to me.

**Comment by martinholters** [comment]
Travis OSX failure was a problem with downloading Julia 0.5. I've seen those recently, but it looks like it started working again shortly after the failure occurred here, in case someone wants to restart the build.

**Comment by mschauer** [comment]
Yes, this was ephemeral.

**Comment by mschauer** [comment]
How about adding parentheses when z is not a single variable?

**Comment by martinholters** [comment]
I certainly wouldn't want to add too much parsing of the variable, the `^` special case already made me uncomfortable. Of course,
```julia
julia> Poly([1, 2, 3], "a+b")
Poly(1 + 2*a+b + 3*a+b^2)
```
does not look good either, but here I would argue the sensible thing to do is to just add the parentheses upon construction, that is
```julia
julia> Poly([1, 2, 3], "(a+b)")
Poly(1 + 2*(a+b) + 3*(a+b)^2)
```
And of course
```
julia> Poly([1, 2, 3], "(z^-1)")
Poly(1 + 2*(z^-1) + 3*(z^-1)^2)
```
is certainly better than without the parentheses, and it's nice that it makes clear what the variable is. But for the use case that motivated me to this PR, `1 + 2*z^-1 + 3*z^-2` is still better: In DSP.jl, `Poly`s are part of a `PolynomialRatio` which I'd like to pretty-print, and here the polynomial evaluation is abstracted away so that the user does not need to know whether negative powers of `z` or positive powers of `z^-1` are used internally.

That said, to make this less specific for one particular use case, I'm considering the alternative of adding an option to provide a custom `printexponent` function to `printpoly`. Opinions?

**Comment by mschauer** [comment]
I would favour adding parentheses if the argument is not a single character variable, defer from all parsing and an option to add or suppress the parentheses.

**Comment by martinholters** [comment]
So the interface should be like this?
```julia
function printpoly{T}(io::IO, p::Poly{T}, mimetype = MIME"text/plain"(); 
                      descending_powers=false, add_parens=length(string(p.var))!=1)
```
I.e. by default (e.g. used by `show`) add parentheses iff variable is not a single character, but allow overriding the default in `printpoly`? Objections against another `varprinter=printexponent` named parameter there?

**Comment by jverzani** [comment]
Boy this gets unwieldy pretty soon. Maybe there could be a symbol type to dispatch on so that these things could be specified by choice of type. It might make also make sense to make this part of `SymbolLike` in the constructor.

**Comment by martinholters** [comment]
But if that choice is made based on the symbol type, the `var` field would have to be of a broader type (or some additional info be stored in an extra field), or am I misunderstanding something?

Come to think of it, a more radical approach that could also solve #77 would be to also store the powers, defaulting to `0:N`, but `N:-1:0` achieving what `descending_powers`what supposed to do and `0:-1:-N` what I'd like for `z^-1`. But that would be a rather fundamental change, when all I wanted was some nicer printing 😄 Oh, and it wouldn't add any parentheses by itself...

**Comment by jverzani** [comment]
Yes, this desire makes me think about two things: should the allowed symbols be more general and should the indexing be more general (without a significant loss in performance). It seems like you want Laurent series to be available. Without looking into it, this could perhaps be achieved without much fuss using an offset array instead of `Vector{T}`. With that and reversing printing you could get most of what you are after without worrying about algebraic simplifications or pretty printing of variables that could be composite. Maybe with that in mind, `printpoly` should have a documented option to reverse order, but `add_parens` the exponent simplification could be an implementation detail that may change going forward.

**Comment by martinholters** [comment]
Hm, yes, an `OffsetArray` for coefficients would suffice for my use case. But for now I'm tempted to just remove anything except the `descending_powers` part from this PR and cope with pretty printing of the powers downstream by a simple `replace`. Anything else would probably require quite some more thought...

Briefly for the background of the negative powers: In DSP, the digital filters in z-transform domain turn out to be polynomial ratios, where the numerator is of lower degree than the denominator, and it is customary to reduce by the highest power of z occurring in the denominator. E.g. instead of

3z / (z² + 2z + 4)

one would write

3z⁻¹ / (1 + 2z⁻¹ + 4z⁻²),

and that is also the natural order to enter the coefficients, i.e. `PolynomialRatio([0, 3], [1, 2, 4])` for this case (although DSP.jl gets that wrong at present), so it makes sense to keep this coefficient order and use z⁻¹ as polynomial variable.

TL;DR: I need _only_ the non-positive powers, not really a full-fledged Laurent series.

**Comment by martinholters** [comment]
So any thoughts on the current state of this PR? Just the possibility to reverse the term order, export, and docs. If this is uncontroversial, can this be merged and other features postponed?

**Comment by jverzani** [comment]
Sorry, missed that you had updated your PR. Thanks!
